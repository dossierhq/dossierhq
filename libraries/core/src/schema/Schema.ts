import { ok, type ErrorType, type Result } from '../ErrorResult.js';
import { BaseSchema } from './BaseSchema.js';
import { PublishedSchema } from './PublishedSchema.js';
import type {
  AdminSchemaMigrationAction,
  AdminSchemaSpecification,
  AdminSchemaSpecificationUpdate,
  AdminSchemaSpecificationWithMigrations,
} from './SchemaSpecification.js';
import { schemaAdminToPublished } from './schemaAdminToPublished.js';
import { schemaUpdate } from './schemaUpdate.js';
import { schemaValidateAdmin } from './schemaValidateAdmin.js';

export class Schema<
  TSpec extends
    | AdminSchemaSpecification
    | AdminSchemaSpecificationWithMigrations = AdminSchemaSpecification,
> extends BaseSchema<TSpec> {
  private cachedPublishedSchema: PublishedSchema | null = null;

  static createAndValidate(
    update: AdminSchemaSpecificationUpdate,
  ): Result<Schema, typeof ErrorType.BadRequest> {
    const result = SchemaWithMigrations.createAndValidate(update);
    if (!result.isOk()) return result;

    const { migrations, ...specWithoutMigrations } = result.value.spec;

    return ok(new Schema(specWithoutMigrations));
  }

  constructor(spec: TSpec) {
    super(spec);
  }

  validate(): Result<void, typeof ErrorType.BadRequest> {
    return schemaValidateAdmin(this);
  }

  toPublishedSchema(): PublishedSchema {
    if (this.cachedPublishedSchema) {
      return this.cachedPublishedSchema;
    }

    const publishedSpec = schemaAdminToPublished(this);

    this.cachedPublishedSchema = new PublishedSchema(publishedSpec);
    return this.cachedPublishedSchema;
  }
}

export class SchemaWithMigrations extends Schema<AdminSchemaSpecificationWithMigrations> {
  static override createAndValidate(
    update: AdminSchemaSpecificationUpdate,
  ): Result<SchemaWithMigrations, typeof ErrorType.BadRequest> {
    const emptySpec: AdminSchemaSpecificationWithMigrations = {
      schemaKind: 'full',
      version: 0,
      entityTypes: [],
      componentTypes: [],
      patterns: [],
      indexes: [],
      migrations: [],
    };
    const empty = new SchemaWithMigrations(emptySpec);
    return empty.updateAndValidate(update);
  }

  updateAndValidate(
    update: AdminSchemaSpecificationUpdate,
  ): Result<SchemaWithMigrations, typeof ErrorType.BadRequest> {
    // Update
    const updatedResult = schemaUpdate(this.spec, update);
    if (updatedResult.isError()) return updatedResult;
    const updatedSpec = updatedResult.value;

    if (updatedSpec === this.spec) {
      return ok(this); // no change
    }

    // Validate
    const updatedSchema = new SchemaWithMigrations(updatedSpec);
    const validateResult = updatedSchema.validate();
    if (validateResult.isError()) return validateResult;

    return ok(updatedSchema);
  }

  collectMigrationActionsSinceVersion(oldSchemaVersion: number): AdminSchemaMigrationAction[] {
    const migrationsToConsider = this.spec.migrations.filter((it) => it.version > oldSchemaVersion);
    if (migrationsToConsider.length === 0) {
      return [];
    }

    migrationsToConsider.sort((a, b) => a.version - b.version);

    const actions = [];
    for (const migration of migrationsToConsider) {
      actions.push(...migration.actions);
    }
    return actions;
  }
}
