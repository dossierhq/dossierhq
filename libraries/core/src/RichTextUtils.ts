import { RichTextNodeType } from './Schema.js';
import type {
  HeadingTagType,
  SerializedHeadingNode,
  SerializedParagraphNode,
  SerializedTextNode,
  TextFormatType,
} from './third-party/Lexical.js';
import { TEXT_TYPE_TO_FORMAT } from './third-party/Lexical.js';
import type {
  EntityReference,
  RichText,
  RichTextEntityLinkNode,
  RichTextEntityNode,
  RichTextNode,
  RichTextValueItemNode,
  ValueItem,
} from './Types.js';

export function createRichTextRootNode(children: RichTextNode[]): RichText {
  return {
    root: {
      direction: 'ltr',
      format: '',
      indent: 0,
      type: 'root',
      version: 1,
      children,
    },
  };
}

export function createRichTextHeadingNode(
  tag: HeadingTagType,
  children: RichTextNode[]
): SerializedHeadingNode {
  return {
    direction: 'ltr',
    format: '',
    indent: 0,
    tag,
    type: 'heading',
    version: 1,
    children,
  };
}

export function createRichTextParagraphNode(children: RichTextNode[]): SerializedParagraphNode {
  return {
    direction: 'ltr',
    format: '',
    indent: 0,
    type: 'paragraph',
    version: 1,
    children,
  };
}

export function createRichTextTextNode(
  text: string,
  options?: { format?: TextFormatType[] }
): SerializedTextNode {
  let formatValue = 0;
  if (options?.format) {
    for (const formatItem of options.format) {
      const formatFlag = TEXT_TYPE_TO_FORMAT[formatItem];
      if (formatFlag !== undefined) {
        formatValue |= formatFlag;
      }
    }
  }
  return {
    detail: 0,
    format: formatValue,
    mode: 'normal',
    style: '',
    text,
    type: 'text',
    version: 1,
  };
}

export function createRichTextEntityNode(reference: EntityReference): RichTextEntityNode {
  return { type: RichTextNodeType.entity, reference, version: 1 };
}

export function createRichTextEntityLinkNode(
  reference: EntityReference,
  children: RichTextNode[]
): RichTextEntityLinkNode {
  return {
    direction: 'ltr',
    format: '',
    indent: 0,
    type: RichTextNodeType.entityLink,
    reference,
    version: 1,
    children,
  };
}

export function createRichTextValueItemNode<T extends ValueItem<string, object>>(
  data: T
): RichTextValueItemNode {
  return { type: RichTextNodeType.valueItem, data, version: 1 };
}
