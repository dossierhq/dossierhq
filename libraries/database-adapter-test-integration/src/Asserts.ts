import type { ErrorType, OkResult, Result } from '@jonasb/datadata-core';
import { createErrorResult, isFieldValueEqual } from '@jonasb/datadata-core';

class AssertionError extends Error {
  actual: unknown;
  expected: unknown;

  constructor(actual: unknown, expected: unknown, message: string) {
    super(message);
    this.name = 'AssertionError';
    this.actual = actual;
    this.expected = expected;
  }
}

export function assertTruthy(actual: unknown): asserts actual {
  if (!actual) {
    throw new AssertionError(actual, true, `Expected truthy`);
  }
}

export function assertSame<T>(actual: T, expected: T): void {
  if (actual !== expected) {
    throw new AssertionError(actual, expected, `Expected same, got ${actual} !== ${expected}`);
  }
}

export function assertNotSame<T>(actual: T, expected: T): void {
  if (actual === expected) {
    throw new AssertionError(actual, expected, `Expected not same, got ${actual} === ${expected}`);
  }
}

export function assertOkResult<TOk, TError extends ErrorType>(
  actual: Result<unknown, ErrorType>
): asserts actual is OkResult<TOk, TError> {
  if (actual.isError()) {
    throw new AssertionError(
      actual,
      undefined,
      `Expected ok, got error ${actual.error}: ${actual.message}`
    );
  }
}

export function assertErrorResult(
  actual: Result<unknown, ErrorType>,
  expectedErrorType: ErrorType,
  expectedMessage: string
): void {
  if (!actual.isError()) {
    throw new AssertionError(
      actual,
      createErrorResult(expectedErrorType, expectedMessage),
      `Expected error, but was ok`
    );
  }
  const expectedString = `${expectedErrorType}: ${expectedMessage}`;
  const actualString = `${actual.error}: ${actual.message}`;
  assertSame(actualString, expectedString);
}

export function assertResultValue<TOk, TError extends ErrorType>(
  result: Result<TOk, TError>,
  expectedValue: TOk
): void {
  assertOkResult(result);
  if (!isFieldValueEqual(result.value, expectedValue)) {
    throw new AssertionError(result.value, expectedValue, `Expected result values to equal`);
  }
}
