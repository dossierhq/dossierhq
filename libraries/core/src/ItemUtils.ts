import { assertExhaustive } from './Asserts.js';
import type { ErrorType, Result } from './ErrorResult.js';
import { notOk, ok } from './ErrorResult.js';
import type {
  AdminFieldSpecification,
  AdminSchema,
  FieldSpecification,
  FieldValueTypeMap,
} from './Schema.js';
import { FieldType, RichTextNodeType } from './Schema.js';
import type {
  AdminEntity,
  AdminEntityCreate,
  AdminEntityUpdate,
  EntityLike,
  PublishedEntity,
  RichText,
  RichTextCodeHighlightNode,
  RichTextCodeNode,
  RichTextElementNode,
  RichTextEntityLinkNode,
  RichTextEntityNode,
  RichTextHeadingNode,
  RichTextLineBreakNode,
  RichTextLinkNode,
  RichTextListItemNode,
  RichTextListNode,
  RichTextNode,
  RichTextParagraphNode,
  RichTextRootNode,
  RichTextTextNode,
  RichTextValueItemNode,
  ValueItem,
} from './Types.js';

/** Check if `value` with `fieldSpec` is a single boolean field */
export function isBooleanField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.Boolean] | null {
  return fieldSpec.type === FieldType.Boolean && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list boolean field */
export function isBooleanListField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is Array<FieldValueTypeMap[typeof FieldType.Boolean]> | null {
  return fieldSpec.type === FieldType.Boolean && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single boolean field or an item in a list field */
export function isBooleanItemField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.Boolean] | null {
  return fieldSpec.type === FieldType.Boolean;
}

/** Check if `value` with `fieldSpec` is a single Entity field */
export function isEntityField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.Entity] | null {
  return fieldSpec.type === FieldType.Entity && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list Entity field */
export function isEntityListField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is Array<FieldValueTypeMap[typeof FieldType.Entity]> | null {
  return fieldSpec.type === FieldType.Entity && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single Entity field or an item in a list field */
export function isEntityItemField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.Entity] | null {
  return fieldSpec.type === FieldType.Entity;
}

/** Check if `value` with `fieldSpec` is a single Location field */
export function isLocationField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.Location] | null {
  return fieldSpec.type === FieldType.Location && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list Location field */
export function isLocationListField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is Array<FieldValueTypeMap[typeof FieldType.Location]> | null {
  return fieldSpec.type === FieldType.Location && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single Location field or an item in a list field */
export function isLocationItemField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.Location] | null {
  return fieldSpec.type === FieldType.Location;
}

/** Check if `value` with `fieldSpec` is a single number field */
export function isNumberField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.Number] | null {
  return fieldSpec.type === FieldType.Number && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list number field */
export function isNumberListField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is Array<FieldValueTypeMap[typeof FieldType.Number]> | null {
  return fieldSpec.type === FieldType.Number && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single number field or an item in a list field */
export function isNumberItemField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.Number] | null {
  return fieldSpec.type === FieldType.Number;
}

/** Check if `value` with `fieldSpec` is a single String field */
export function isStringField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.String] | null {
  return fieldSpec.type === FieldType.String && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list String field */
export function isStringListField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is Array<FieldValueTypeMap[typeof FieldType.String]> | null {
  return fieldSpec.type === FieldType.String && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single String field or an item in a list field */
export function isStringItemField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.String] | null {
  return fieldSpec.type === FieldType.String;
}

/** Check if `value` with `fieldSpec` is a single RichText field */
export function isRichTextField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.RichText] | null {
  return fieldSpec.type === FieldType.RichText && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list RichText field */
export function isRichTextListField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is Array<FieldValueTypeMap[typeof FieldType.RichText]> | null {
  return fieldSpec.type === FieldType.RichText && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single RichText field or an item in a list field */
export function isRichTextItemField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.RichText] | null {
  return fieldSpec.type === FieldType.RichText;
}

/** Check if `value` with `fieldSpec` is a single ValueItem field */
export function isValueItemField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.ValueItem] | null {
  return fieldSpec.type === FieldType.ValueItem && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list ValueItem field */
export function isValueItemListField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is Array<FieldValueTypeMap[typeof FieldType.ValueItem]> | null {
  return fieldSpec.type === FieldType.ValueItem && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single ValueItem field or an item in a list field */
export function isValueItemItemField(
  fieldSpec: FieldSpecification,
  value: unknown | null
): value is FieldValueTypeMap[typeof FieldType.ValueItem] | null {
  return fieldSpec.type === FieldType.ValueItem;
}

export function isRichTextTextNode(node: RichTextNode): node is RichTextTextNode {
  return node.type === RichTextNodeType.text;
}

export function isRichTextLineBreakNode(node: RichTextNode): node is RichTextLineBreakNode {
  return node.type === RichTextNodeType.linebreak;
}

export function isRichTextElementNode(node: RichTextNode): node is RichTextElementNode {
  return 'children' in node;
}

export function isRichTextRootNode(node: RichTextNode): node is RichTextRootNode {
  return node.type === RichTextNodeType.root;
}

export function isRichTextParagraphNode(node: RichTextNode): node is RichTextParagraphNode {
  return node.type === RichTextNodeType.paragraph;
}

export function isRichTextEntityNode(node: RichTextNode): node is RichTextEntityNode {
  return node.type === RichTextNodeType.entity;
}

export function isRichTextEntityLinkNode(node: RichTextNode): node is RichTextEntityLinkNode {
  return node.type === RichTextNodeType.entityLink;
}

export function isRichTextValueItemNode(node: RichTextNode): node is RichTextValueItemNode {
  return node.type === RichTextNodeType.valueItem;
}

export function isRichTextCodeNode(node: RichTextNode): node is RichTextCodeNode {
  return node.type === RichTextNodeType.code;
}

export function isRichTextCodeHighlightNode(node: RichTextNode): node is RichTextCodeHighlightNode {
  return node.type === RichTextNodeType['code-highlight'];
}

export function isRichTextHeadingNode(node: RichTextNode): node is RichTextHeadingNode {
  return node.type === RichTextNodeType.heading;
}

export function isRichTextLinkNode(node: RichTextNode): node is RichTextLinkNode {
  return node.type === RichTextNodeType.link;
}

export function isRichTextListNode(node: RichTextNode): node is RichTextListNode {
  return node.type === RichTextNodeType.list;
}

export function isRichTextListItemNode(node: RichTextNode): node is RichTextListItemNode {
  return node.type === RichTextNodeType.listitem;
}

export function isItemValueItem(
  item:
    | ValueItem
    | PublishedEntity
    | AdminEntity
    | AdminEntityCreate
    | AdminEntityUpdate
    | EntityLike
): item is ValueItem {
  return 'type' in item;
}

export function isItemAdminEntity(
  item: ValueItem | PublishedEntity | AdminEntity
): item is AdminEntity {
  return !isItemValueItem(item) && 'version' in item.info;
}

export function isItemEntity(
  item: ValueItem | PublishedEntity | AdminEntity
): item is PublishedEntity {
  return !isItemValueItem(item) && !isItemAdminEntity(item);
}

export type ItemValuePath = (string | number)[];

export function visitorPathToString(path: ItemValuePath): string {
  let result = '';
  for (const segment of path) {
    if (Number.isInteger(segment)) {
      result += `[${segment}]`;
    } else {
      if (result.length === 0) {
        result += segment;
      } else {
        result += `.${segment}`;
      }
    }
  }
  return result;
}

export function copyEntity<
  T extends
    | AdminEntity<string, object>
    | AdminEntityCreate<AdminEntity<string, object>>
    | PublishedEntity<string, object>
>(
  entity: T,
  changes: { id?: string; info?: Partial<T['info']>; fields?: Partial<T['fields']> }
): T {
  const copy = { ...entity };
  if (typeof changes.id === 'string') {
    copy.id = changes.id;
  }
  if (changes.info) {
    copy.info = { ...entity.info };
    for (const [key, value] of Object.entries(changes.info)) {
      (copy.info as unknown as Record<string, unknown>)[key] = value;
    }
  }
  if (changes.fields) {
    const fieldsCopy: Record<string, unknown> = { ...entity.fields };
    copy.fields = fieldsCopy;
    for (const [key, value] of Object.entries(changes.fields)) {
      fieldsCopy[key] = value;
    }
  }
  return copy;
}

export function isEntityNameAsRequested(currentName: string, requestedName: string): boolean {
  if (requestedName === currentName) {
    return true;
  }
  const hashIndex = currentName.lastIndexOf('#');
  if (hashIndex < 0) {
    return false;
  }
  const currentWithoutUniqueNumber = currentName.slice(0, hashIndex);
  return requestedName === currentWithoutUniqueNumber;
}

export function isFieldValueEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;

  if (a === null || a === undefined || b === null || b === undefined) {
    return false; // if a or be are not defined they can't be equal
  }

  if (Array.isArray(a)) {
    if (!Array.isArray(b)) return false;
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i += 1) {
      if (!isFieldValueEqual(a[i], b[i])) {
        return false;
      }
    }
    return true;
  }

  if (typeof a === 'object') {
    if (typeof b !== 'object' || a === null || b === null) return false;
    const aKeys = Object.keys(a);
    const bKeys = Object.keys(b);
    if (aKeys.length !== bKeys.length) return false;
    for (const key of aKeys) {
      if (
        !isFieldValueEqual((a as Record<string, unknown>)[key], (b as Record<string, unknown>)[key])
      ) {
        return false;
      }
    }
    return true;
  }

  return false;
}

export function normalizeEntityFields<TEntity extends EntityLike>(
  schema: AdminSchema,
  entity: EntityLike,
  options?: { excludeOmitted: boolean }
): Result<TEntity['fields'], typeof ErrorType.BadRequest> {
  const entitySpec = schema.getEntityTypeSpecification(entity.info.type);
  if (!entitySpec) {
    return notOk.BadRequest(`Entity type ${entity.info.type} doesn’t exist`);
  }

  const unsupportedFieldNames = new Set(Object.keys(entity.fields));

  const fields: Record<string, unknown> = {};
  for (const fieldSpec of entitySpec.fields) {
    unsupportedFieldNames.delete(fieldSpec.name);

    if (options?.excludeOmitted && (!entity.fields || !(fieldSpec.name in entity.fields))) {
      continue;
    }

    const fieldValue = normalizeFieldValue(
      schema,
      fieldSpec,
      entity.fields?.[fieldSpec.name] ?? null
    );
    fields[fieldSpec.name] = fieldValue;
  }

  if (unsupportedFieldNames.size > 0) {
    return notOk.BadRequest(`Unsupported field names: ${[...unsupportedFieldNames].join(', ')}`);
  }

  return ok(fields);
}

export function normalizeValueItem<TValueItem extends ValueItem>(
  schema: AdminSchema,
  valueItem: TValueItem
): Result<TValueItem, typeof ErrorType.BadRequest> {
  const valueTypeSpec = schema.getValueTypeSpecification(valueItem.type);
  if (!valueTypeSpec) {
    return notOk.BadRequest(`Value type ${valueItem.type} doesn’t exist`);
  }

  const unsupportedFieldNames = new Set(Object.keys(valueItem));
  unsupportedFieldNames.delete('type');

  const result: ValueItem = { type: valueItem.type };
  for (const fieldSpec of valueTypeSpec.fields) {
    unsupportedFieldNames.delete(fieldSpec.name);

    if (fieldSpec.name === 'type') {
      continue;
    }

    const fieldValue = normalizeFieldValue(schema, fieldSpec, valueItem[fieldSpec.name] ?? null);
    result[fieldSpec.name] = fieldValue;
  }

  if (unsupportedFieldNames.size > 0) {
    return notOk.BadRequest(`Unsupported field names: ${[...unsupportedFieldNames].join(', ')}`);
  }

  return ok(result as TValueItem);
}

export function normalizeFieldValue(
  schema: AdminSchema,
  fieldSpec: AdminFieldSpecification,
  value: unknown
): unknown {
  if (value === null || value === undefined) {
    return null;
  }

  if (fieldSpec.list) {
    if (!Array.isArray(value)) {
      return value; // Invalid
    }

    let changed = false;
    const newList = [];
    for (let i = 0; i < value.length; i += 1) {
      const item = value[i];
      const normalizedItem = normalizeFieldValueItem(schema, fieldSpec, item);
      if (item !== normalizedItem || normalizedItem === null) {
        changed = true;
      }
      if (normalizedItem !== null) {
        newList.push(normalizedItem);
      }
    }

    if (newList.length === 0) {
      return null;
    }
    return changed ? newList : value;
  }

  return normalizeFieldValueItem(schema, fieldSpec, value);
}

function normalizeFieldValueItem(
  schema: AdminSchema,
  fieldSpec: AdminFieldSpecification,
  value: unknown
) {
  if (value === null) return value;

  const type = fieldSpec.type as FieldType;
  switch (type) {
    case FieldType.Boolean:
    case FieldType.Entity:
    case FieldType.Location:
    case FieldType.Number:
      return value;
    case FieldType.RichText: {
      // TODO normalize ValueItem nodes?
      // Don't fail on invalid rich text, just ignore it in this phase
      const richText = value as RichText;
      if (
        typeof richText === 'object' &&
        typeof richText.root === 'object' &&
        Array.isArray(richText.root.children) &&
        richText.root.children.length === 1
      ) {
        const onlyChild = richText.root.children[0];
        if (
          onlyChild.type === RichTextNodeType.paragraph &&
          isRichTextElementNode(onlyChild) &&
          onlyChild.children.length === 0
        ) {
          return null;
        }
      }
      return value;
    }
    case FieldType.String:
      //TODO support trimming of strings?
      return value ? value : null;
    case FieldType.ValueItem: {
      const valueItem = value as ValueItem;
      const newValueItem: ValueItem = { type: valueItem.type };
      let changed = false;

      const valueSpec = schema.getValueTypeSpecification(valueItem.type);
      if (!valueSpec) {
        return value; // Invalid
      }
      for (const fieldFieldSpec of valueSpec.fields) {
        const fieldName = fieldFieldSpec.name;
        const fieldValue = valueItem[fieldName];
        const normalizedFieldValue =
          normalizeFieldValue(schema, fieldFieldSpec, fieldValue) ?? null;
        newValueItem[fieldName] = normalizedFieldValue;
        if (normalizedFieldValue !== fieldValue) {
          changed = true;
        }
      }

      for (const [fieldName, fieldValue] of Object.entries(valueItem)) {
        if (!(fieldName in newValueItem)) {
          // Invalid, so just reuse initial value
          newValueItem[fieldName] = fieldValue;
        }
      }

      return changed ? newValueItem : valueItem;
    }
    default: {
      assertExhaustive(type);
    }
  }
}
