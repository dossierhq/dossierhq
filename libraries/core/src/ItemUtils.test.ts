import { describe, expect, test } from 'vitest';
import { assertIsDefined } from './Asserts.js';
import { isEntityNameAsRequested, isFieldValueEqual, normalizeFieldValue } from './ItemUtils.js';
import { createRichTextParagraphNode, createRichTextRootNode } from './RichTextUtils.js';
import { AdminSchema, FieldType } from './Schema.js';

const schema = new AdminSchema({
  entityTypes: [
    {
      name: 'Foo',
      adminOnly: false,
      authKeyPattern: null,
      fields: [
        { name: 'string', type: FieldType.String },
        { name: 'stringList', type: FieldType.String, list: true },
        { name: 'twoStrings', type: FieldType.ValueType, valueTypes: ['TwoStrings'] },
        { name: 'richText', type: FieldType.RichText },
      ],
    },
  ],
  valueTypes: [
    {
      name: 'TwoStrings',
      adminOnly: false,
      fields: [
        { name: 'string1', type: FieldType.String },
        { name: 'string2', type: FieldType.String },
      ],
    },
  ],
  patterns: [],
});

function getEntityFieldSpec(schema: AdminSchema, entityType: string, fieldName: string) {
  const entitySpec = schema.getEntityTypeSpecification(entityType);
  assertIsDefined(entitySpec);
  const fieldSpec = schema.getEntityFieldSpecification(entitySpec, fieldName);
  assertIsDefined(fieldSpec);
  return fieldSpec;
}

describe('isEntityNameAsRequested', () => {
  test('hello=hello', () => expect(isEntityNameAsRequested('hello', 'hello')).toBeTruthy());
  test('hello#123=hello', () => expect(isEntityNameAsRequested('hello#123', 'hello')).toBeTruthy());
  test('hello#123=hello#123', () =>
    expect(isEntityNameAsRequested('hello#123', 'hello#123')).toBeTruthy());

  test('hello!=world', () => expect(isEntityNameAsRequested('hello', 'world')).toBeFalsy());
  test('hello#456!=hello#123', () =>
    expect(isEntityNameAsRequested('hello#456', 'hello#123')).toBeFalsy());
});

describe('isFieldValueEqual', () => {
  test('string===string', () => expect(isFieldValueEqual('hello', 'hello')).toBeTruthy());
  test('string!==null', () => expect(isFieldValueEqual('hello', null)).toBeFalsy());
  test('string!==other string', () => expect(isFieldValueEqual('hello', 'world')).toBeFalsy());
  test('string[]===string[]', () =>
    expect(isFieldValueEqual(['hello', 'world'], ['hello', 'world'])).toBeTruthy());
  test('string[]!==string[] (order)', () =>
    expect(isFieldValueEqual(['hello', 'world'], ['world', 'hello'])).toBeFalsy());

  test('value item===value item', () =>
    expect(
      isFieldValueEqual(
        {
          type: 'Foo',
          string: 'string',
          stringList: ['string', 'list'],
          entity: { id: 'entity-id-1' },
          entityList: [{ id: 'entity-id-1' }],
        },
        {
          type: 'Foo',
          string: 'string',
          stringList: ['string', 'list'],
          entity: { id: 'entity-id-1' },
          entityList: [{ id: 'entity-id-1' }],
        }
      )
    ).toBeTruthy());

  test('value item!==value item', () =>
    expect(
      isFieldValueEqual(
        {
          type: 'Foo',
          string: 'string',
          stringList: ['string', 'list'],
          entity: { id: 'entity-id-1' },
          entityList: [{ id: 'entity-id-1' }],
        },
        {
          type: 'Foo',
          string: 'string',
          stringList: ['string', 'DIFFERENCE'],
          entity: { id: 'entity-id-1' },
          entityList: [{ id: 'entity-id-1' }],
        }
      )
    ).toBeFalsy());
});

describe('normalizeFieldValue()', () => {
  test('"" => null', () =>
    expect(normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'string'), '')).toEqual(
      null
    ));

  test('[] => null', () =>
    expect(normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'stringList'), [])).toBe(
      null
    ));

  test('[string, ""] => [string]', () =>
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'stringList'), ['hello', ''])
    ).toEqual(['hello']));

  test('[string] => [string] (no change)', () => {
    const fieldValue = ['hello', 'world'];
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'stringList'), fieldValue)
    ).toBe(fieldValue);
  });

  test('{string1:string,string2:""} => {string1:string,string2:null}', () =>
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'twoStrings'), {
        type: 'TwoStrings',
        string1: 'Hello',
        string2: '',
      })
    ).toEqual({ type: 'TwoStrings', string1: 'Hello', string2: null }));

  test('{string1:undefined} => {string1:null,string2:null}', () =>
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'twoStrings'), {
        type: 'TwoStrings',
        string1: undefined,
        // no string2
      })
    ).toEqual({ type: 'TwoStrings', string1: null, string2: null }));

  test('{string1:string,string2:string} => {string1:string,string2:string} (no change)', () => {
    const fieldValue = {
      type: 'TwoStrings',
      string1: 'Hello',
      string2: 'World',
    };
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'twoStrings'), fieldValue)
    ).toBe(fieldValue);
  });

  test('string undefined => undefined', () =>
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'string'), undefined)
    ).toBe(undefined));

  test('string[] undefined => undefined', () =>
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'stringList'), undefined)
    ).toBe(undefined));

  test('ValueItem: undefined => undefined', () => {
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'twoStrings'), undefined)
    ).toBe(undefined);
  });

  test('RichText: empty paragraph => null', () => {
    expect(
      normalizeFieldValue(
        schema,
        getEntityFieldSpec(schema, 'Foo', 'richText'),
        createRichTextRootNode([createRichTextParagraphNode([])])
      )
    ).toBe(null);
  });

  test('RichText: let invalid rich text pass through (no root)', () => {
    expect(normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'richText'), {})).toEqual(
      {}
    );
  });

  test('RichText: let invalid rich text pass through (string in root)', () => {
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'richText'), {
        root: 'hello world',
      })
    ).toEqual({ root: 'hello world' });
  });
});
