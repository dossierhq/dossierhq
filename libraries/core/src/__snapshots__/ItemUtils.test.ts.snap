// Vitest Snapshot v1

exports[`normalizeEntityFields > empty Foo 1`] = `
{
  "richText": null,
  "string": null,
  "stringList": null,
  "twoStrings": null,
}
`;

exports[`normalizeValueItem > TwoStrings with empty strings 1`] = `
{
  "string1": null,
  "string2": null,
  "type": "TwoStrings",
}
`;

exports[`normalizeValueItem > empty TwoStrings 1`] = `
{
  "string1": null,
  "string2": null,
  "type": "TwoStrings",
}
`;
